gobblin-data-management/src/main/java/gobblin/data/management/copy/CopySource.java

```diff
@@ -200,13 +210,11 @@ private FileSystem getOptionallyThrottledFileSystem(FileSystem fs, State state)
 
       try {
 
-        Collection<? extends CopyEntity> files = this.copyableDataset.getCopyableFiles(this.targetFs, this.copyConfiguration);
-        List<FileSet<CopyEntity>> fileSets = partitionCopyableFiles(this.copyableDataset, files);
-
-        // Sort to optimize the insertion to work units list
-        Collections.sort(fileSets, this.workUnitList.getComparator());
+        Iterator<FileSet<CopyEntity>> fileSets =
+            this.copyableDataset.getFileSetIterator(this.targetFs, this.copyConfiguration);
 
-        for (FileSet<CopyEntity> fileSet : fileSets) {
+        while (fileSets.hasNext() && !this.workUnitList.hasRejectedFileSet()) {
```

Comment:
* `hasRejectedFileSet` only returns `true` with an actual rejection (false even if work units have been evicted). It is just a way of letting caller know whether it is likely that the new work units will also be rejected. There could be a different method to check if anything has been evicted.
* There is no issue if it returns false and then it becomes true after the work units are created. The new work units will be rejected or accepted in the same way. This is just a really dump heuristic to know when to stop generating work units.